### PR TITLE
Add common module

### PR DESCRIPTION
**Pull request by chimp1984**
Based on https://github.com/bisq-network/misq/pull/3

I did only add few tests so far. 
If any developer wants to contribute to add more unit tests to that module would be great. I need to focus to get progress on other areas...Some tests can be taken over from Bisq as well.

### CODE REVIEW

File: common/src/main/java/network/misq/common/currency/FiatCurrency.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+import java.util.Currency;
+import java.util.Locale;
+
+@EqualsAndHashCode(callSuper = true)
+@ToString
+
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/currency/FiatCurrency.java]:
```suggestion
```

File: common/src/main/java/network/misq/common/currency/MisqCurrency.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    @EqualsAndHashCode.Exclude
+    protected final String name;
+
+    public static boolean isFiat(String code) {
+        return FiatCurrencyRepository.getFiatCurrencyByCode().containsKey(code);
+    }
+
+    // We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
+    // of crypto currencies to be flexible with any newly added one. 
+    public static boolean isMaybeCrypto(String code) {
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/currency/MisqCurrency.java]: Use JavaDoc for this comment
```suggestion
    /**
     * We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
     * of crypto currencies to be flexible with any newly added one.
     */
    public static boolean isMaybeCrypto(String code) {
```

File: common/src/main/java/network/misq/common/locale/LocaleRepository.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+public class LocaleRepository {
+    // To be set early at application setup
+    @Setter
+    @Getter
+    private static Locale defaultLocale;
+
+    static {
+        defaultLocale = Locale.getDefault();
+        // On some systems there is no country defined, in that case we use en_US
+        if (defaultLocale.getCountry() == null || defaultLocale.getCountry().isEmpty()) {
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/locale/LocaleRepository.java:35]: If you're into that 
```suggestion
        if (CollectionUtils.isEmpty(defaultLocale.getCountry())) {
```

File: common/src/main/java/network/misq/common/locale/LocaleRepository.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            new Locale("en", "UG"), // Uganda / lang=English
+            new Locale("uk", "UA"), // Ukraine / lang=Ukrainian
+            new Locale("ar", "AE"), // United Arab Emirates / lang=Arabic
+            new Locale("en", "GB"), // United Kingdom of Great Britain and Northern Ireland / lang=English
+            new Locale("en", "US"), // United States of America / lang=English
+            new Locale("es", "UY"), // Uruguay / lang=Spanish
+            new Locale("uz", "UZ"), // Uzbekistan / lang=Uzbek
+            new Locale("bi", "VU"), // Vanuatu / lang=Bislama
+            new Locale("es", "VE"), // Venezuela (Bolivarian Republic of) / lang=Spanish
+            new Locale("vi", "VN"), // Viet Nam / lang=Vietnamese
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/locale/LocaleRepository.java]:
```suggestion
            new Locale("vi", "VN"), // Vietnam / lang=Vietnamese
```

File: common/src/main/java/network/misq/common/locale/Region.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        REGION_CODE_BY_COUNTRY_CODE.put("UA", "EU"); // name=Ukraine / region=Europe / subregion=Eastern Europe
+        REGION_CODE_BY_COUNTRY_CODE.put("AE", "AS"); // name=United Arab Emirates / region=Asia / subregion=Western Asia
+        REGION_CODE_BY_COUNTRY_CODE.put("GB", "EU"); // name=United Kingdom of Great Britain and Northern Ireland / region=Europe / subregion=Northern Europe
+        REGION_CODE_BY_COUNTRY_CODE.put("US", "AM"); // name=United States of America / region=Americas / subregion=Northern America
+        REGION_CODE_BY_COUNTRY_CODE.put("UY", "AM"); // name=Uruguay / region=Americas / subregion=South America
+        REGION_CODE_BY_COUNTRY_CODE.put("UZ", "AS"); // name=Uzbekistan / region=Asia / subregion=Central Asia
+        REGION_CODE_BY_COUNTRY_CODE.put("VU", "OC"); // name=Vanuatu / region=Oceania / subregion=Melanesia
+        REGION_CODE_BY_COUNTRY_CODE.put("VE", "AM"); // name=Venezuela (Bolivarian Republic of) / region=Americas / subregion=South America
+        REGION_CODE_BY_COUNTRY_CODE.put("VN", "AS"); // name=Viet Nam / region=Asia / subregion=South-Eastern Asia
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/locale/Region.java]:
```suggestion
        REGION_CODE_BY_COUNTRY_CODE.put("VN", "AS"); // name=Vietnam / region=Asia / subregion=South-Eastern Asia
```

File: common/src/main/java/network/misq/common/locale/Region.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        REGION_CODE_BY_COUNTRY_CODE.put("ZW", "AF"); // name=Zimbabwe / region=Africa / subregion=Eastern Africa
+    }
+
+    public static String getRegionName(final String regionCode) {
+        return NAME_BY_CODE.get(regionCode);
+    }
+
+    public static String getRegionCode(String countryCode) {
+        return REGION_CODE_BY_COUNTRY_CODE.getOrDefault(countryCode, "Undefined");
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/locale/Region.java:291]: Why does one method return a default value and not the other?

File: common/src/main/java/network/misq/common/monetary/Coin.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    public static Coin parseXmr(String string) {
+        return Coin.of(new BigDecimal(string).movePointRight(12).longValue(),
+                "XMR",
+                12);
+    }
+
+    // Unsafe method. We do not know the currency and the smallestUnitExponent, just that it is not a fiat currency.
+    // We exclude XMR
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Coin.java:55]: Use JavaDoc for this comment

File: common/src/main/java/network/misq/common/monetary/Coin.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    public static Coin of(long value, String currencyCode) {
+        return new Coin(value, currencyCode, deriveExponent(currencyCode));
+    }
+
+    public static Coin of(double value, String currencyCode) {
+        int exponent = deriveExponent(currencyCode);
+        return new Coin(value, currencyCode, exponent);
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Coin.java:91]: Inline the return statement like in the Coin.of() above this one

File: common/src/main/java/network/misq/common/monetary/Coin.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        return new Coin(LongMath.checkedSubtract(this.value, value.value), this.currencyCode, this.smallestUnitExponent);
+    }
+
+    public Coin multiply(long factor) {
+        return new Coin(LongMath.checkedMultiply(this.value, factor), this.currencyCode, this.smallestUnitExponent);
+    }
+
+    public Coin divide(long divisor) {
+        return new Coin(this.value / divisor, this.currencyCode, this.smallestUnitExponent);
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Coin.java:122]: Check zero division?

File: common/src/main/java/network/misq/common/monetary/Coin.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    private static int deriveExponent(String currencyCode) {
+        return currencyCode.equals("XMR") ? 12 : 8;
+    }
+
+    @Override
+    public String toString() {
+        return "Coin{" +
+                "\r\n} " + super.toString();
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Coin.java:137]: It should be inside the curly braces, right?
```suggestion
    public String toString() {
        return "Coin{" + super.toString() + "\r\n} ";
    }
```

File: common/src/main/java/network/misq/common/monetary/Coin.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        return new Coin(value, currencyCode, smallestUnitExponent);
+    }
+
+    Coin(long value, String currencyCode, int smallestUnitExponent) {
+        super(value, currencyCode, smallestUnitExponent);
+    }
+
+    private Coin(double value, String currencyCode, int smallestUnitExponent) {
+        super(value, currencyCode, smallestUnitExponent);
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Coin.java:104]: Constructors above methods by convention

File: common/src/main/java/network/misq/common/monetary/Fiat.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        return new Fiat(value, currencyCode, smallestUnitExponent);
+    }
+
+    Fiat(long value, String currencyCode, int smallestUnitExponent) {
+        super(value, currencyCode, smallestUnitExponent);
+    }
+
+    private Fiat(double value, String currencyCode, int smallestUnitExponent) {
+        super(value, currencyCode, smallestUnitExponent);
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Fiat.java:73]: Constructors above methods by convention

File: common/src/main/java/network/misq/common/monetary/Fiat.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        return new Fiat(LongMath.checkedSubtract(this.value, value.value), this.currencyCode, this.smallestUnitExponent);
+    }
+
+    public Fiat multiply(long factor) {
+        return new Fiat(LongMath.checkedMultiply(this.value, factor), this.currencyCode, this.smallestUnitExponent);
+    }
+
+    public Fiat divide(long divisor) {
+        return new Fiat(this.value / divisor, this.currencyCode, this.smallestUnitExponent);
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Fiat.java:91]: Check zero division?

File: common/src/main/java/network/misq/common/monetary/Fiat.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    @Override
+    public double toDouble(long value) {
+        return MathUtils.roundDouble(BigDecimal.valueOf(value).movePointLeft(smallestUnitExponent).doubleValue(), smallestUnitExponent);
+    }
+
+    @Override
+    public String toString() {
+        return "Fiat{" +
+                "\r\n} " + super.toString();
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Fiat.java:102]:
```suggestion
    public String toString() {
        return "Fiat{" + super.toString() + "\r\n} ";
    }
```

File: common/src/main/java/network/misq/common/monetary/Monetary.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    protected final long value;
+    protected final String currencyCode;
+    protected final int smallestUnitExponent;
+
+    public static Monetary from(Monetary monetary, long newValue) {
+        if (monetary instanceof Fiat) {
+            return Fiat.of(newValue, monetary.getCurrencyCode(), monetary.getSmallestUnitExponent());
+        } else {
+            return Coin.of(newValue, monetary.getCurrencyCode(), monetary.getSmallestUnitExponent());
+        }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Monetary.java:40]: Redundant `else` clause
```suggestion
        if (monetary instanceof Fiat) {
            return Fiat.of(newValue, monetary.getCurrencyCode(), monetary.getSmallestUnitExponent());
        } 
        return Coin.of(newValue, monetary.getCurrencyCode(), monetary.getSmallestUnitExponent());
```

File: common/src/main/java/network/misq/common/monetary/Monetary.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    public static Monetary from(long amount, String currencyCode) {
+        if (MisqCurrency.isFiat(currencyCode)) {
+            return Fiat.of(amount, currencyCode);
+        } else {
+            return Coin.of(amount, currencyCode);
+        }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Monetary.java:48]: Redundant `else` clause

File: common/src/main/java/network/misq/common/monetary/Monetary.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    protected Monetary(long value, String currencyCode, int smallestUnitExponent) {
+        this.value = value;
+        this.smallestUnitExponent = smallestUnitExponent;
+        this.currencyCode = currencyCode;
+    }
+
+    protected Monetary(double value, String currencyCode, int smallestUnitExponent) {
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Monetary.java:57]: Constructors above methods by convention

File: common/src/main/java/network/misq/common/monetary/Quote.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+                "baseMonetary must be the same type as the quote.baseMonetary");
+        Monetary quoteMonetary = quote.quoteMonetary;
+        long value = BigDecimal.valueOf(baseMonetary.value).multiply(BigDecimal.valueOf(quote.value))
+                .movePointLeft(baseMonetary.smallestUnitExponent)
+                .longValue();
+        if (quoteMonetary instanceof Fiat) {
+            return new Fiat(value,
+                    quoteMonetary.currencyCode,
+                    quoteMonetary.smallestUnitExponent);
+        } else {
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/monetary/Quote.java:151]: redundant `else` clause

File: common/src/main/java/network/misq/common/util/ByteArrayUtils.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        for (byte[] aByte : bytes) {
+            byteBuffer.put(aByte);
+        }
+        return byteBuffer.array();
+    }
+
+    public static byte[] integerToByteArray(int intValue, int numBytes) {
+        byte[] bytes = new byte[numBytes];
+        for (int i = numBytes - 1; i >= 0; i--) {
+            bytes[i] = ((byte) (intValue & 0xFF));
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/ByteArrayUtils.java:39]: This AND should be redundant bc the byte cast takes the last 8 bits anyway
```suggestion
            bytes[i] = (byte) intValue;
```

File: common/src/main/java/network/misq/common/util/ByteArrayUtils.java [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+            intValue >>>= 8;
+        }
+        return bytes;
+    }
+
+    public static int byteArrayToInteger(byte[] bytes) {
+        int result = 0;
+        for (byte aByte : bytes) {
+            result = result << 8 | aByte & 0xff;
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/ByteArrayUtils.java:48]:
```suggestion
            result = result << 8 | aByte;
```

File: common/src/main/java/network/misq/common/util/DecimalFormatters.java [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+        public final String format(long number) {
+            return decimalFormat.format(number);
+        }
+
+        public final String format(double number) {
+            return decimalFormat.format(number);
+        }
+    }
+
+    // Caches formatters which have the same parameters. We use formatters as stateless immutable objects.
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/DecimalFormatters.java]: Use JavaDoc comment

File: common/src/main/java/network/misq/common/util/DecimalFormatters.java [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    private static final LoadingCache<Couple<Locale, Integer>, Format> decimalFormatCache = CacheBuilder.newBuilder()
+            .build(CacheLoader.from(DecimalFormatters::getDecimalFormat));
+
+    // Returns cached DecimalFormat object.
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/DecimalFormatters.java]: Use JavaDoc comment

File: common/src/main/java/network/misq/common/util/FileUtils.java [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+        if (!file.exists() && !file.createNewFile()) {
+            throw new IOException("Could not make file " + file);
+        }
+    }
+
+    public static void writeToFile(String string, File file) throws IOException {
+        try (FileWriter fileWriter = new FileWriter(file.getAbsolutePath())) {
+            fileWriter.write(string);
+        } catch (IOException e) {
+            log.warn("Could not {} to file {}", string, file);
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/FileUtils.java]:
```suggestion
            log.warn("Could not write {} to file {}", string, file);
```

File: common/src/main/java/network/misq/common/util/MapUtils.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+public class MapUtils {
+    public static void disposeAndRemove(String key, Map<String, ? extends Disposable> map) {
+        if (map.containsKey(key)) {
+            map.get(key).dispose();
+            map.remove(key);
+        }
+    }
+
+    public static void disposeAndRemoveAll(Map<String, ? extends Disposable> map) {
+        map.keySet().forEach(key -> disposeAndRemove(key, map));
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/MapUtils.java]: More efficient
```suggestion
        map.keySet().forEach(Disposable::dispose);
        map.clear();
```

File: common/src/main/java/network/misq/common/util/OsUtils.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+            return arch.endsWith("64")
+                    || wow64Arch != null && wow64Arch.endsWith("64")
+                    ? "64" : "32";
+        } else if (osArch.contains("arm")) {
+            // armv8 is 64 bit, armv7l is 32 bit
+            return osArch.contains("64") || osArch.contains("v8") ? "64" : "32";
+        } else if (isLinux()) {
+            return osArch.startsWith("i") ? "32" : "64";
+        } else {
+            return osArch.contains("64") ? "64" : osArch;
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/OsUtils.java]: Redundant `else` clauses

File: common/src/main/java/network/misq/common/util/StatisticsUtils.java [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+        int middle = list.length / 2;
+        long median;
+        if (list.length % 2 == 1) {
+            median = list[middle];
+        } else {
+            median = MathUtils.roundDoubleToLong((list[middle - 1] + list[middle]) / 2.0);
+        }
+        return median;
+    }
```
Comment (kaiwolfram) [inline common/src/main/java/network/misq/common/util/StatisticsUtils.java]: Only works if the array is already sorted

File: common/src/main/java/network/misq/common/util/OsUtils.java (hunk 25 above)
Comment (chimp1984) [reply common/src/main/java/network/misq/common/util/OsUtils.java]: For my feeling its better readability with leaving it in the if/else branches.

Comment (kaiwolfram) [reply common/src/main/java/network/misq/common/util/OsUtils.java]: Everyone has personal preferences. I only mentioned it bc it's in the style guide https://github.com/bisq-network/style/issues/12

File: common/src/main/java/network/misq/common/locale/LocaleRepository.java (hunk 3 above)
Comment (chimp1984) [reply common/src/main/java/network/misq/common/locale/LocaleRepository.java:35]: Ah I have to revert that commit as we do not have apache dependency yet and as long there is not a strong reason don't plan to add it.

File: common/src/main/java/network/misq/common/util/StatisticsUtils.java (hunk 26 above)
Comment (chimp1984) [reply common/src/main/java/network/misq/common/util/StatisticsUtils.java]: I rename the list and add a javadoc.

**Comment by chimp1984** [comment]
@KaiWitt Great thanks for the reviews!

File: common/src/main/java/network/misq/common/util/OsUtils.java (hunk 25 above)
Comment (chimp1984) [reply common/src/main/java/network/misq/common/util/OsUtils.java]: Ah true. You know the style guide better than myself ;-). But then I will break up the if/else branches as well...

File: common/src/main/java/network/misq/common/util/MapUtils.java [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+public class MapUtils {
+    public static void disposeAndRemove(String key, Map<String, ? extends Disposable> map) {
+        if (map.containsKey(key)) {
+            map.get(key).dispose();
+            map.remove(key);
+        }
+    }
+
+    public static void disposeAndRemoveAll(Map<String, ? extends Disposable> map) {
+        map.keySet().forEach(Disposable::dispose);
```
Comment (ghost) [inline common/src/main/java/network/misq/common/util/MapUtils.java]: This did not build on my system.  What did seem to work instead was:
```suggestion
        map.forEach((k,v) -> v.dispose());
```

**Review by jmacxx — CHANGES_REQUESTED** [review]
Other than the one build error, there were several complaints from git about trailing spaces.

```
../misqpull4.patch:480: trailing whitespace.
    
../misqpull4.patch:788: trailing whitespace.
    // We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
../misqpull4.patch:789: trailing whitespace.
    // of crypto currencies to be flexible with any newly added one. 
../misqpull4.patch:3868: trailing whitespace.
     * We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
../misqpull4.patch:239: new blank line at EOF.
+
warning: 4 lines applied after fixing whitespace errors.
```

File: common/src/main/java/network/misq/common/util/MapUtils.java (hunk 27 above)
Comment (chimp1984) [reply common/src/main/java/network/misq/common/util/MapUtils.java]: Ah accepted the suggestion without building it ;-(
Should have been:
`map.values().forEach(Disposable::dispose);`

**Comment by chimp1984** [comment]
> Other than the one build error, there were several complaints from git about trailing spaces.
> 
> ```
> ../misqpull4.patch:480: trailing whitespace.
>     
> ../misqpull4.patch:788: trailing whitespace.
>     // We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
> ../misqpull4.patch:789: trailing whitespace.
>     // of crypto currencies to be flexible with any newly added one. 
> ../misqpull4.patch:3868: trailing whitespace.
>      * We only can check if the currency is not fiat and if the code matches the format, but we do not maintain a list 
> ../misqpull4.patch:239: new blank line at EOF.
> +
> warning: 4 lines applied after fixing whitespace errors.
> ```

How did you get those complaints?

**Comment by jmacxx** [comment]
The whitespace errors are produced when applying the PR locally using `git apply`.

We should set up some automatic build tools like the Bisq repo has, perhaps can discuss at the status meeting.

**Comment by chimp1984** [comment]
> The whitespace errors are produced when applying the PR locally using `git apply`.
> 
> We should set up some automatic build tools like the Bisq repo has, perhaps can discuss at the status meeting.

Yes definitely! Should have been done initially, forgot about it...

**Review by jmacxx — APPROVED** [review]
utACK

**Review by jmacxx — APPROVED** [review]
utACK

**Comment by ghubstan** [comment]
ACK

**Pull request merged**